### PR TITLE
Updating `ivy.put_along_axis`

### PR DESCRIPTION
**Pull request by mobley-trent**
This PR converts `ivy.put_along_axis` into a mixed function with paddle and torch backends. I then go on to implement the frontend implementations for torch and paddle that make use of this augmented version.

### CODE REVIEW

**Comment by ivy-leaves** [comment]
If you are working on an open task, please **edit** the PR description to link to the issue you've created.

For more information, please check [ToDo List Issues Guide](https://unify.ai/docs/ivy/overview/contributing/the_basics.html#todo-list-issues).

Thank you :hugs:

**Review by AnnaTz — CHANGES_REQUESTED** [review]

File: ivy/functional/ivy/experimental/manipulation.py
Comment (mobley-trent) [reply ivy/functional/ivy/experimental/manipulation.py:1788]: So the whole function should be written from the ground up ?

Comment (AnnaTz) [reply ivy/functional/ivy/experimental/manipulation.py:1788]: I don't think the function needs to be re-written altogether. Just the handling of `mode` and `out`.

**Comment by vedpatwardhan** [comment]
Hey @mobley-trent, as far as writing a partial mixed function is concerned, everything seems to be in place with the decorator, the handlers and the wrapper attribute. The failures in the tests are actually problems with the implementation itself, for e.g. there's a failure (here)[https://github.com/unifyai/ivy/actions/runs/5600955176/jobs/10244117918?pr=18809] where we're trying to concat arrays of [incompatible](https://github.com/unifyai/ivy/actions/runs/5600955176/jobs/10244117918?pr=18809#step:3:319) shapes along with other issues which @AnnaTz mentioned before. Thanks @mobley-trent 😄👍

File: ivy/functional/frontends/torch/tensor_functions.py
Comment (jkeane508) [inline ivy/functional/frontends/torch/tensor_functions.py:37]: We should probably include `scatter_add_` and `scatter_reduce_`. These are both placeholder functions but there may be cases where these are called by the GC.

**Review by jkeane508 — CHANGES_REQUESTED** [review]

File: ivy/functional/ivy/experimental/manipulation.py
Comment (mobley-trent) [reply ivy/functional/ivy/experimental/manipulation.py:1847]: Numpy doesn't support the reduction operation natively and I'm finding it a bit hard to implement. I could use the same algorithm that I used in `ivy.put_along_axis` but won't that be just code repetition ?

Comment (AnnaTz) [reply ivy/functional/ivy/experimental/manipulation.py:1847]: We could only use the native numpy function for the `assign` mode then.

**Comment by mobley-trent** [comment]
> I'm getting this error trace while trying to run the test :
> 
> ```
> ========================================================================= short test summary info =========================================================================
> ERROR ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_manipulation.py - ivy.utils.exceptions.IvyAttributeError: paddle: function_supported_devices_and_dtypes: paddle: function_supported_devices_and_dtypes: 'tuple' object has no attribute 'values'
> ============================================================================ 1 error in 5.92s =============================================================================
> ERROR: not found: /workspaces/ivy/ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_manipulation.py::test_put_along_axis
> (no name '/workspaces/ivy/ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_manipulation.py::test_put_along_axis' in any of [<Module test_manipulation.py>])
> ```
> 
> I'm also getting the same error trace when trying to run the example in the script I've written in `scratch.py`. However, when I run the function as `ivy.put_along_axis` instead of `put_along_axis`, it runs okay. @sherry30 do you think you can help with this ?

Nvm I just pulled from master and I'm not getting this issue any more 👍

File: ivy/functional/ivy/experimental/manipulation.py
Comment (mobley-trent) [reply ivy/functional/ivy/experimental/manipulation.py:1855]: Should `np.put_along_axis` frontend be implemented ? @AnnaTz

Comment (AnnaTz) [reply ivy/functional/ivy/experimental/manipulation.py:1862]: Eventually yes, but it's not necessary to do so in this PR.

File: ivy/data_classes/container/experimental/manipulation.py
Comment (AnnaTz) [inline ivy/data_classes/container/experimental/manipulation.py:3685]: We are supposed to add the `key_chains`, `map_sequences` etc. arguments when defining the container methods.

File: ivy/functional/backends/numpy/experimental/manipulation.py
Comment (AnnaTz) [inline ivy/functional/backends/numpy/experimental/manipulation.py:521]: Needs support for `out`.

File: ivy/functional/backends/paddle/manipulation.py
Comment (AnnaTz) [inline ivy/functional/backends/paddle/manipulation.py:507]: Likewise needs support for `out`.

File: ivy/functional/ivy/experimental/manipulation.py
Comment (AnnaTz) [inline ivy/functional/ivy/experimental/manipulation.py:1778]: `Optional` should only be used when the default value is `None`. Same goes for the array and container method definitions above.

File: ivy/functional/ivy/experimental/manipulation.py
Comment (AnnaTz) [inline ivy/functional/ivy/experimental/manipulation.py:1828]: I think we can remove this comment.

File: ivy_tests/test_ivy/test_frontends/test_paddle/test_tensor/test_manipulation.py
Comment (AnnaTz) [inline ivy_tests/test_ivy/test_frontends/test_paddle/test_tensor/test_manipulation.py:655]: Why did you choose to set `max_axis=1` here? I don't think we need to limit the test like this.

File: ivy_tests/test_ivy/test_frontends/test_paddle/test_tensor/test_manipulation.py
Comment (AnnaTz) [inline ivy_tests/test_ivy/test_frontends/test_paddle/test_tensor/test_manipulation.py:660]: I think we can remove this since you have set `min_axis=0` above.

File: ivy_tests/test_ivy/test_frontends/test_paddle/test_tensor/test_manipulation.py
Comment (AnnaTz) [inline ivy_tests/test_ivy/test_frontends/test_paddle/test_tensor/test_manipulation.py:671]: I think we should be using `shape` here, not `idx_shape`.

File: ivy_tests/test_ivy/test_frontends/test_paddle/test_tensor/test_manipulation.py
Comment (AnnaTz) [inline ivy_tests/test_ivy/test_frontends/test_paddle/test_tensor/test_manipulation.py:666]: I think the values at each dimension of the indices should be at most the length of the corresponding dimension. Maybe you could have a look at `test_scatter_nd` to get an idea on how to generate this.

File: ivy_tests/test_ivy/test_frontends/test_torch/test_tensor.py
Comment (AnnaTz) [inline ivy_tests/test_ivy/test_frontends/test_torch/test_tensor.py:11902]: I don't think we should be hardcoding this to `int64`. Same for the `test_tensor_functions` and the experimental api helper.

File: ivy_tests/test_ivy/test_frontends/test_torch/test_tensor.py
Comment (AnnaTz) [inline ivy_tests/test_ivy/test_frontends/test_torch/test_tensor.py:11929]: Same comments as for the paddle test. 
I think since the ivy test and all the frontend tests are so similar we could have one single helper for all of them.

**Review by AnnaTz — CHANGES_REQUESTED** [review]
Great PR! Just a few minor corrections in the functions and some improvements in the testing needed.

File: ivy/functional/ivy/experimental/manipulation.py
Comment (AnnaTz) [reply ivy/functional/ivy/experimental/manipulation.py:1862]: Actually on second thought, yes, please add the frontend functions for `numpy.put_along_axis` and `jax.numpy.put_along_axis` as well. Should be very quick once the rest are up and working with their tests.

Comment (mobley-trent) [reply ivy/functional/ivy/experimental/manipulation.py:1862]: `jax.numpy` does not have `put_along_axis`

Comment (AnnaTz) [reply ivy/functional/ivy/experimental/manipulation.py:1862]: Oh my bad, it only has `apply_along_axis`. Just the numpy one then. Thanks!

File: ivy_tests/test_ivy/test_frontends/test_paddle/test_tensor/test_manipulation.py
Comment (mobley-trent) [reply ivy_tests/test_ivy/test_frontends/test_paddle/test_tensor/test_manipulation.py:660]: It was drawing negative values during some test runs

Comment (AnnaTz) [reply ivy_tests/test_ivy/test_frontends/test_paddle/test_tensor/test_manipulation.py:660]: Could you add a ToDo comment there to keep track of the problem then? Something like "`dtype_values_axis` currently returns negative axis even for min_axis=0. Remove this condition when the issue is fixed."

File: ivy/functional/ivy/experimental/manipulation.py
Comment (mobley-trent) [inline ivy/functional/ivy/experimental/manipulation.py:1846]: @AnnaTz for some inputs I have to get the transpose at this stage i.e `ivy.vstack(ret).T`, but I don't know what is causing this issue. This was the main pain point in this PR. Otherwise everything else is working fine

File: ivy/functional/ivy/experimental/manipulation.py
Comment (AnnaTz) [inline ivy/functional/ivy/experimental/manipulation.py:1847]: I think the problem with the needed `.T` has to do with the order in which you are accumulating results in `ret`.  Instead of appending individual results and then stacking them and reshaping at the end, you could try something like this:
```
ret = ivy.zeros_like(arr)

for ii in _ndindex(Ni):
    for kk in _ndindex(Nk):
        arr_1d = arr[ii + (slice(None),) + kk]
        indices_1d = indices[ii + (slice(None),) + kk]
        values_1d = values[ii + (slice(None),) + kk]
        ret[ii + (slice(None),) + kk] = _handle_reduction_op(mode, arr_1d, indices_1d, values_1d)
```

Comment (AnnaTz) [reply ivy/functional/ivy/experimental/manipulation.py:1847]: However, ideally we would want to avoid `for` loops altogether since they are very inefficient.

**Review by AnnaTz — CHANGES_REQUESTED** [review]

File: ivy_tests/test_ivy/test_frontends/test_torch/test_tensor.py
Comment (mobley-trent) [reply ivy_tests/test_ivy/test_frontends/test_torch/test_tensor.py:12253]: It only works with `int64` because it casts this dtype to the rest of the arguments in method input dtypes. The index tensor has to be of type `int64`.

Comment (AnnaTz) [reply ivy_tests/test_ivy/test_frontends/test_torch/test_tensor.py:12253]: You can pass lists of dtypes to `test_frontend_method`. `init_input_dtypes` would be this `input_dtype` which can be any of supported types, `method_input_dtypes` would be [`int64`, int64`, `input_dtype`] (if values has to be the same dtype as `x`, otherwise would be another `val_dtype` which `put_along_axis_helper` would have to return as well).

Comment (mobley-trent) [reply ivy_tests/test_ivy/test_frontends/test_torch/test_tensor.py:12253]: Also for the test for `ivy.put_along_axis` the test is failing unless the available dtypes in the helper is set to `["int64"]`, probably caused by the same issue here.

The tests for the scatter instance methods because of this error 
```
ivy.utils.exceptions.IvyBackendException: torch: put_along_axis: Index tensor must have the same number of dimensions as self tensor
```
with this set of examples:
```
args=('int64', array([[0, 0],  [0, 0]]), array([[0, 0]]), array([[0, 0]]), 0)
```

These are the values being produced by the strategy. They run fine when you test them yourself with the function in a script somewhere but they are failing the testing pipeline somehow.

I also set the max value inside the index tensor to be either 0 or 1, because I couldn't find a way to draw only values that would return an index error when performing the slicing operations. The mechanism works fine though, its only that sometimes it could draw values which were greater than the length of `arr[indices]` in  the `_handle_reduction_op` helper function.

File: ivy/functional/frontends/torch/tensor.py
Comment (mobley-trent) [reply ivy/functional/frontends/torch/tensor.py:1669]: Like that ?

Comment (AnnaTz) [reply ivy/functional/frontends/torch/tensor.py:1671]: Can just be 
```
        self.ivy_array = ivy.put_along_axis(self.ivy_array, index, src, dim, mode=reduce)
 ```
 ivy functions already return ivy arrays

File: ivy_tests/test_ivy/test_frontends/test_torch/test_tensor.py
Comment (AnnaTz) [reply ivy_tests/test_ivy/test_frontends/test_torch/test_tensor.py:12253]: Hi @mobley-trent, I tried what I suggested it to you and `test_torch_scatter` passes for all backends except paddle. Please do that to solve the `int64` issue.

Comment (AnnaTz) [reply ivy_tests/test_ivy/test_frontends/test_torch/test_tensor.py]: Does `init_input_dtypes = [input_dtype]` and `method_input_dtypes = ["int64", input_dtype]` work for the methods?

Comment (AnnaTz) [reply ivy_tests/test_ivy/test_frontends/test_torch/test_tensor.py]: Basically `init_input_dtypes` should provide the dtypes for all arrays contained in `init_all_as_kwargs_np`, and likewise `method_input_dtypes` should provide the dtypes for all arrays contained in `method_all_as_kwargs_np`.

File: scratch.py
Comment (AnnaTz) [inline scratch.py:101]: Let's not forget to remove this extra file before merging 😅

**Review by AnnaTz — CHANGES_REQUESTED** [review]

File: ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_manipulation.py
Comment (AnnaTz) [reply ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_manipulation.py:1194]: Then you need to set the `ground_truth_backend`. `backend_to_test` defines which backend you are testing against the ground truth.

File: ivy/functional/backends/numpy/experimental/manipulation.py
Comment (AnnaTz) [inline ivy/functional/backends/numpy/experimental/manipulation.py:511]: Hi @mobley-trent, `test_put_along_axis` is passing now 💪
I just forgot to update the numpy and paddle backend with the new `mode` names, so currently only the torch backend is being used. Please update the rest and fix any issues that might come up.

Comment (AnnaTz) [inline ivy/functional/backends/numpy/experimental/manipulation.py:511]: Oh and the `mode` typehints should be the same as in the main implementation. I updated it in torch but it also needs to change in numpy and paddle.

Comment (AnnaTz) [inline ivy/functional/backends/numpy/experimental/manipulation.py:511]: And I didn't look into the frontends at all. You'll have to run them and see if something needs to be changed. 
`ivy.put_along_axis` accumulates in case there are duplicate indices, so if a frontend is supposed to not do that, we'll have to clear up `indices` before calling the ivy function.

(For the missing modes in the torch frontend you can just add a ToDo.)

**Review by AnnaTz — CHANGES_REQUESTED** [review]

File: ivy/functional/ivy/experimental/manipulation.py
Comment (AnnaTz) [inline ivy/functional/ivy/experimental/manipulation.py:1876]: I probably shouldn't have used `copy_array` here now that I think about it.

**Review by AnnaTz — CHANGES_REQUESTED** [review]

File: ivy/functional/ivy/experimental/manipulation.py
Comment (AnnaTz) [reply ivy/functional/ivy/experimental/manipulation.py:1876]: The way the `out` argument works in ivy functions is: If `out` is None the return is a new array, else the array provided by `out` is in-place updated with the return.
The backends and the compositional implementation need to be updated to do exactly that.

File: ivy/functional/backends/numpy/experimental/manipulation.py
Comment (mobley-trent) [reply ivy/functional/backends/numpy/experimental/manipulation.py:511]: I think torch supports all the modes natively

Comment (AnnaTz) [reply ivy/functional/backends/numpy/experimental/manipulation.py:511]: Yes, I meant the `mean` and `mul` modes which will not be supported yet because they are missing from our `scatter_nd`.

Comment (mobley-trent) [reply ivy/functional/backends/numpy/experimental/manipulation.py:511]: Numpy only supports `replace` and Paddle doesn't support `mean`, `max`, or `mean`. The function is a partial mixed function, so the function jumps directly to the implemented `put_along_axis` backends without running `scatter_nd` in the compositional implementation

**Review by AnnaTz — CHANGES_REQUESTED** [review]

File: ivy/functional/backends/jax/general.py
Comment (mobley-trent) [reply ivy/functional/backends/jax/general.py:391]: The mode for `mul` won't work without these modes being supported in the `scatter_nd` backends for jax, numpy and tensorflow. As the function is partially mixed, there were some inconsistencies since the `put_along_axis` backends for torch and paddle support these modes, while the compositional implementation doesn't. I was thinking we can at least add the `mul` mode for now and `mean` can be implemented in the future. What do you think ?

Comment (AnnaTz) [reply ivy/functional/backends/jax/general.py:391]: `scatter_nd` needs to be updated across all backends to support `mean` and `mul`. There is no use in updating only `jax` and `numpy`. So, we will exclude `mul` and `mean` from `put_along_axis` for now. There will be no inconsistency as we will not be testing for these anywhere.

File: ivy/functional/backends/tensorflow/general.py
Comment (AnnaTz) [inline ivy/functional/backends/tensorflow/general.py:399]: If this approach works correctly, I think you can do something similar for "mean" as well 🤔
Maybe even for the torch and paddle backends... 
If `scatter_nd` works correctly like this we can include all the modes for `put_along_axis` after all.

**Review by AnnaTz — CHANGES_REQUESTED** [review]

**Comment by AnnaTz** [comment]
ivy-gardener

**Comment by AnnaTz** [comment]
@mobley-trent I have addressed the remaining unresolved review comments myself. Please next time read them more carefully before marking them as completed.

**Comment by mobley-trent** [comment]
> @mobley-trent I have addressed the remaining unresolved review comments myself. Please next time read them more carefully before marking them as completed.

Sorry @AnnaTz I thought it was in relation to our discussion on Discord 🙏

**Comment by AnnaTz** [comment]
> > @mobley-trent I have addressed the remaining unresolved review comments myself. Please next time read them more carefully before marking them as completed.
> 
> Sorry @AnnaTz I thought it was in relation to our discussion on Discord 🙏

No worries. Now I'm just waiting for the CI checks to finish. If things pass, we can finally merge this!

**Review by AnnaTz — APPROVED** [review]

**Pull request merged**
